registry/registry_pubkey_update.go

```diff
@@ -0,0 +1,359 @@
+		"server_payload":   hex.EncodeToString(serverPayload),
+		"server_signature": hex.EncodeToString(serverSignature),
+	}
+	return res, nil
+}
+
+// Update the public key of registered prefix(es) if the http request passed client and server verification for nonce.
+// It returns the response data, and an error if any
+func updateNsKeySignChallengeCommit(ctx *gin.Context, data *RegisteredPrefixUpdate) (map[string]interface{}, error) {
+	// Validate the client's jwks as a set here
+	key, err := validateJwks(string(data.Pubkey))
+	if err != nil {
+		return nil, badRequestError{Message: err.Error()}
+	}
+	var rawkey interface{} // This is the raw key, like *rsa.PrivateKey or *ecdsa.PrivateKey
+	if err := key.Raw(&rawkey); err != nil {
+		return nil, errors.Wrap(err, "failed to generate raw pubkey from jwks")
+	}
+
+	// Verify the Proof of Possession of the client and server's active private keys
+	clientPayload := []byte(data.ClientNonce + data.ServerNonce)
+	clientSignature, err := hex.DecodeString(data.ClientSignature)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode the client's signature")
+	}
+	clientVerified := verifySignature(clientPayload, clientSignature, (rawkey).(*ecdsa.PublicKey))
+	serverPayload, err := hex.DecodeString(data.ServerPayload)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode the server's payload")
+	}
+
+	serverSignature, err := hex.DecodeString(data.ServerSignature)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode the server's signature")
+	}
+
+	serverPrivateKey, err := loadServerKeys()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to decode the server's private key")
+	}
+	serverPubkey := serverPrivateKey.PublicKey
+	serverVerified := verifySignature(serverPayload, serverSignature, &serverPubkey)
+
+	if clientVerified && serverVerified {
+		for _, prefix := range data.Prefixes {
+			log.Debug("Start updating namespace: ", prefix)
+
+			// Check if prefix exists before doing anything else
+			exists, err := namespaceExistsByPrefix(prefix)
+			if err != nil {
+				log.Errorf("Failed to check if namespace already exists: %v", err)
+				return nil, errors.Wrap(err, "Server encountered an error checking if namespace already exists")
+			}
+			if exists {
+				// Update the namespace's public key with the latest one when authorized origin provides a new key
+				existingNs, err := getNamespaceByPrefix(prefix)
+				if err != nil {
+					log.Errorf("Failed to get existing namespace to update: %v", err)
+					return nil, errors.Wrap(err, "Server encountered an error getting existing namespace to update")
+				}
+
+				// Check the origin is authorized to update (possessing the public key used for prefix initial registration)
+				// Parse all public keys of the sender into a JWKS
+				var clientKeySet jwk.Set
+				if data.AllPubkeys == nil { // backward compatibility - AllPubkeys only exists in the payload in Pelican 7.12 or later
```

Comment:
I'm confused by this.

`data` is from the client -- why is there backward compatibility code in an API call introduced in 7.12?